Pal/security/linux/graphene-sandbox.c

```diff
@@ -0,0 +1,1037 @@
+#include <linux/socket.h>
+#include <linux/un.h>
+#include <linux/net.h>
+#include <linux/atomic.h>
+#include <net/sock.h>
+#include <net/inet_sock.h>
+#include <net/tcp_states.h>
+
+#include "graphene-sandbox.h"
+#include "util.h"
+
+enum {
+	OP_OPEN,
+	OP_STAT,
+	OP_BIND,
+	OP_CONNECT,
+	OP_SENDMSG,
+	OP_RECVMSG,
+};
+
+static atomic_t sandbox_ids = ATOMIC_INIT(0);
+
+static void drop_graphene_info(struct graphene_info *info)
+{
+	struct graphene_path *p, *n;
+
+	list_for_each_entry_safe(p, n, &info->gi_paths, list) {
+		__putname(p->path->name);
+		kfree(p->path);
+		kfree(p);
+	}
+
+	list_for_each_entry_safe(p, n, &info->gi_rpaths, list) {
+		__putname(p->path->name);
+		kfree(p->path);
+		kfree(p);
+	}
+
+	if (info->gi_loader_name) {
+		__putname(info->gi_loader_name->name);
+		kfree(info->gi_loader_name);
+	}
+
+	if (info->gi_mcast_sock)
+		fput(info->gi_mcast_sock);
+
+	kfree(info);
+}
+
+void put_graphene_info(struct graphene_info *info)
+{
+	if (!atomic_dec_return(&info->gi_count))
+		drop_graphene_info(info);
+}
+
+static int check_path(struct graphene_path *gp,
+		      const char *path, int len, u32 mask,
+		      int is_recursive)
+{
+	if (unlikely(!gp->path_len)) {
+		if (path[0] != '/')
+			goto matched;
+		return 0;
+	}
+
+	if (gp->path_len == len) {
+		if (!strncmp(gp->path->name, path, len))
+			goto matched;
+	} else if (is_recursive &&
+		   gp->path_len < len) {
+		if (!strncmp(gp->path->name, path, gp->path_len) &&
+		    path[gp->path_len] == '/')
+			goto matched;
+	}
+
+	return 0;
+
+matched:
+	if (mask & (MAY_READ|MAY_EXEC|MAY_ACCESS)) {
+		if (!(gp->type & GRAPHENE_FS_READ))
+			return -EPERM;
+	}
+
+	if (mask & (MAY_WRITE|MAY_APPEND)) {
+		if (!(gp->type & GRAPHENE_FS_WRITE))
+			return -EPERM;
+	}
+
+	return 1;
+}
+
+static int __common_file_perm(struct graphene_info *gi,
+			      const char *path, u32 mask)
+{
+	struct graphene_path *p;
+	int len = strlen(path);
+	int rv = 0;
+
+	list_for_each_entry(p, &gi->gi_paths, list) {
+		rv = check_path(p, path, len, mask, false);
+		if (rv)
+			goto out;
+	}
+
+	list_for_each_entry(p, &gi->gi_rpaths, list) {
+		rv = check_path(p, path, len, mask, true);
+		if (rv)
+			goto out;
+	}
+
+	rv = -EPERM;
+out:
+	if (rv >= 0) {
+		rv = 0;
+#ifdef GRAPHENE_DEBUG
+		printk(KERN_INFO "Graphene: PID %d PATH %s PASSED\n",
+		       current->pid, path);
+#endif
+	} else {
+		printk(KERN_INFO "Graphene: PID %d PATH %s DENIED\n",
+		       current->pid, path);
+	}
+	return rv;
+}
+
+int check_open_path(struct graphene_info *gi, const char *path,
+		    int flags)
+{
+	return __common_file_perm(gi, path, ACC_MODE(flags));
+}
+
+int check_stat_path(struct graphene_info *gi, const char *path)
+{
+	return __common_file_perm(gi, path, MAY_ACCESS);
+}
+
+int __unix_perm(struct graphene_info *gi,
+		struct sockaddr *address, int addrlen)
+{
+	const char * sun_path =
+		((struct sockaddr_un *) address)->sun_path;
+
+	if (!memcmp(sun_path, gi->gi_unix, GRAPHENE_UNIX_PREFIX_SIZE)) {
```

Comment:
Do you have a guarantee that `sun_path` has at least `GRAPHENE_UNIX_PREFIX_SIZE` bytes? Otherwise it may cause OOB read.